account/account.go

```diff
@@ -41,64 +36,319 @@ func SignBytes(chainID string, o Signable) []byte {
 	buf, n, err := new(bytes.Buffer), new(int), new(error)
 	o.WriteSignBytes(chainID, buf, n, err)
 	if *err != nil {
-		sanity.PanicCrisis(err)
+		panic(fmt.Sprintf("could not write sign bytes for a signable: %s", *err))
 	}
-
 	return buf.Bytes()
 }
 
-//-----------------------------------------------------------------------------
+type Addressable interface {
+	// Get the 20 byte EVM address of this account
+	Address() Address
+	// Public key from which the Address is derived
+	PublicKey() PublicKey
+}
 
-// Account resides in the application state, and is mutated by transactions
-// on the blockchain.
-// Serialized by wire.[read|write]Reflect
-type Account struct {
-	Address     []byte        `json:"address"`
-	PubKey      crypto.PubKey `json:"pub_key"`
-	Sequence    int           `json:"sequence"`
-	Balance     int64         `json:"balance"`
-	Code        []byte        `json:"code"`         // VM code
-	StorageRoot []byte        `json:"storage_root"` // VM storage merkle root.
+// The default immutable interface to an account
+type Account interface {
+	Addressable
+	// The value held by this account in terms of the chain-native token
+	Balance() uint64
+	// The EVM byte code held by this account (or equivalently, this contract)
+	Code() Bytecode
+	// The sequence number (or nonce) of this account, incremented each time a mutation of the
+	// Account is persisted to the blockchain state
+	Sequence() uint64
```

Comment:
It's something of a cascade from having block heights as `uint64`, see my issue here: https://github.com/tendermint/tendermint/issues/731. `int` only takes us to 68 years! Clearly nonce can grow at faster linear rate to this, so makes sense to pull in.

Reply:
It's only adds overhead on systems that have 32-bit words, which IOT possibilities aside, is probably not that many. On 64-bit systems it's free.

Also I do not think it's unnecessary - suppose I hammer a chain using the same account as input with 100 txs per block - unusual but perfectly within the realm of load testing - then we run out of int32s in about 8 months. Worse than this is _if_ we have validators running on a mix of 32-bit and 64-bit architectures the 32-bit ones will die while the 64-bit ones won't. By forcing this as uint64 then we have parity across architectures - and the overhead is really minimal, just space not time really because we do minimal arithmetic on these. So I think this make sense right now, even if we don't _need_ it for height (even then - simulating a long-running chain - the disparity between archs is not nice)